### PR TITLE
Fix issues with GeometryTest address identified by address sanitizer 

### LINKED ISSUE #17438: Geometry::IndexingUtils: AddressSanitizer: stack-buffer-overflow
From a run of units tests with Address Sanitizer (compiled with `-DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_FLAGS=-fsanitize=address`):

```
207/1683 Test #207: GeometryTest_IndexingUtilsTest ...............................***Failed    2.04 sec
Running 27 tests..=================================================================
==83617==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7fff4f7bff00 at pc 0x7f78aef6f3f8 bp 0x7fff4f7bf270 sp 0x7fff4f7bf260
READ of size 8 at 0x7fff4f7bff00 thread T0
    #0 0x7f78aef6f3f7 in Mantid::Geometry::IndexingUtils::GetFirstMaxIndex(double const*, unsigned long, double) /home/simon/code/mantid/Framework/Geometry/src/Crystal/IndexingUtils.cpp:1453
    #1 0x7f78aef6df45 in Mantid::Geometry::IndexingUtils::FFTScanFor_Directions(std::vector<Mantid::Kernel::V3D, std::allocator<Mantid::Kernel::V3D> >&, std::vector<Mantid::Kernel::V3D, std::allocator<Mantid::Kernel::V3D> > const&, double, double, double, double) /home/simon/code/mantid/Framework/Geometry/src/Crystal/IndexingUtils.cpp:1273
    #2 0x7f78aef690f7 in Mantid::Geometry::IndexingUtils::Find_UB(Mantid::Kernel::Matrix<double>&, std::vector<Mantid::Kernel::V3D, std::allocator<Mantid::Kernel::V3D> > const&, double, double, double, double) /home/simon/code/mantid/Framework/Geometry/src/Crystal/IndexingUtils.cpp:494
    #3 0x944813 in IndexingUtilsTest::test_Find_UB_using_FFT() /home/simon/code/mantid/Framework/Geometry/test/IndexingUtilsTest.h:203
    #4 0x94dcd3 in TestDescription_suite_IndexingUtilsTest_test_Find_UB_using_FFT::runTest() /home/simon/build/mantid-debug/Framework/Geometry/test/IndexingUtilsTest.cpp:38
    #5 0xaa32af in CxxTest::RealTestDescription::run() /home/simon/code/mantid/Testing/Tools/cxxtest/cxxtest/RealDescriptions.cpp:96
    #6 0xaa9284 in CxxTest::TestRunner::runTest(CxxTest::TestDescription&) /home/simon/code/mantid/Testing/Tools/cxxtest/cxxtest/TestRunner.h:94
    #7 0xaa8ea5 in CxxTest::TestRunner::runSuite(CxxTest::SuiteDescription&) /home/simon/code/mantid/Testing/Tools/cxxtest/cxxtest/TestRunner.h:79
    #8 0xaa8a75 in CxxTest::TestRunner::runWorld() /home/simon/code/mantid/Testing/Tools/cxxtest/cxxtest/TestRunner.h:63
    #9 0xaa86c0 in CxxTest::TestRunner::runAllTests(CxxTest::TestListener&) /home/simon/code/mantid/Testing/Tools/cxxtest/cxxtest/TestRunner.h:24
    #10 0xab63f9 in CxxTest::XUnitPrinter::run() /home/simon/code/mantid/Testing/Tools/cxxtest/cxxtest/XUnitPrinter.h:30
    #11 0xab7ee1 in int CxxTest::Main<CxxTest::XUnitPrinter>(CxxTest::XUnitPrinter&, int, char**) /home/simon/code/mantid/Testing/Tools/cxxtest/cxxtest/TestMain.h:106
    #12 0xaa15c7 in main /home/simon/build/mantid-debug/Framework/Geometry/test/GeometryTest_runner.cpp:27
    #13 0x7f78abf22a3f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x20a3f)
    #14 0x477568 in _start (/home/simon/build/mantid-debug/bin/GeometryTest+0x477568)

Address 0x7fff4f7bff00 is located in stack of thread T0 at offset 2720 in frame
    #0 0x7f78aef6d6ed in Mantid::Geometry::IndexingUtils::FFTScanFor_Directions(std::vector<Mantid::Kernel::V3D, std::allocator<Mantid::Kernel::V3D> >&, std::vector<Mantid::Kernel::V3D, std::allocator<Mantid::Kernel::V3D> > const&, double, double, double, double) /home/simon/code/mantid/Framework/Geometry/src/Crystal/IndexingUtils.cpp:1193

  This frame has 12 object(s):
    [32, 40) 'fit_error'
    [96, 120) 'full_list'
    [160, 184) 'max_fft_val'
    [224, 248) 'max_fft_copy'
    [288, 312) 'temp_dirs'
    [352, 376) 'temp'
    [416, 440) 'temp_dirs_2'
    [480, 504) 'current_dir'
    [544, 568) 'index_vals'
    [608, 632) 'indexed_qs'
    [672, 2720) 'magnitude_fft' <== Memory access at offset 2720 overflows this variable
    [2752, 6848) 'projections'
HINT: this may be a false positive if your program uses some custom stack unwind mechanism or swapcontext
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /home/simon/code/mantid/Framework/Geometry/src/Crystal/IndexingUtils.cpp:1453 Mantid::Geometry::IndexingUtils::GetFirstMaxIndex(double const*, unsigned long, double)
Shadow bytes around the buggy address:
  0x100069eeff90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100069eeffa0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100069eeffb0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100069eeffc0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100069eeffd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x100069eeffe0:[f2]f2 f2 f2 00 00 00 00 00 00 00 00 00 00 00 00
  0x100069eefff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100069ef0000: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100069ef0010: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100069ef0020: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100069ef0030: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Contiguous container OOB:fc
  ASan internal:           fe
```

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work.**

**Report to:** [nobody]. 

**To test:**

Code review

Fixes #17438

*This does not require release notes* because **it is too small to require it**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Geometry/test/CSGObjectTest.h [hunk 1]
```diff
@@ -934,7 +935,7 @@ class CSGObjectTest : public CxxTest::TestSuite {
     TS_ASSERT_DELTA(cuboid->volume(), cylinderVolume, 1e-6)
   }
 
-  void testMonteCarloVolume() {
+  void xtestMonteCarloVolume() {
```
Comment (SimonHeybrock) [inline Framework/Geometry/test/CSGObjectTest.h]: Disabled?

File: Framework/Geometry/src/Crystal/IndexingUtils.cpp [hunk 2]
```diff
@@ -1474,7 +1473,9 @@ double IndexingUtils::GetFirstMaxIndex(const double magnitude_fft[], size_t N,
   if (found_max) {
     double sum = 0;
     double w_sum = 0;
-    for (size_t j = i - 2; j <= i + 2; j++) {
+    // only include single extra index if we're too near the end
+    const size_t last_index = i < N - 2 ? i + 2 : i + 1;
+    for (size_t j = i - 2; j <= last_index; j++) {
```
Comment (SimonHeybrock) [inline Framework/Geometry/src/Crystal/IndexingUtils.cpp]: Is this easier to read when written as `for (size_t j = i - 2; j < std::min(N, i + 3) ; j++)`?

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]

File: Framework/Geometry/test/CSGObjectTest.h (hunk 1 above)
Comment (martyngigg) [reply Framework/Geometry/test/CSGObjectTest.h]: Nice catch!

File: Framework/Geometry/src/Crystal/IndexingUtils.cpp (hunk 2 above)
Comment (martyngigg) [reply Framework/Geometry/src/Crystal/IndexingUtils.cpp]: That is nice, I like that. Thanks.

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**